File: leetcode/hard/214_Shortest_Palindrome.py
"""
You are given a string s. You can convert s to a
palindrome
by adding characters in front of it.

Return the shortest palindrome you can find by performing this transformation.

 

Example 1:

Input: s = "aacecaaa"
Output: "aaacecaaa"

Example 2:

Input: s = "abcd"
Output: "dcbabcd"

 

Constraints:

    0 <= s.length <= 5 * 104
    s consists of lowercase English letters only.
"""
class Solution:
    def shortestPalindrome(self, s: str) -> str:
        prefix = 0
        suffix = 0
        base = 29
        last_index = 0
        power = 1
        mod = 10**9 + 7

        for i, c in enumerate(s):
            char = (ord(c) - ord('a') + 1)

            prefix = (prefix * base) % mod
            prefix = (prefix + char) % mod
            suffix = (suffix + char * power) % mod
            power = (power * base) % mod

            if prefix == suffix:
                last_index = i
        suffix = s[last_index + 1:]
        return suffix[::-1] + s

File: leetcode/hard/214_Shortest_Palindrome.go
func reverse(input string) string {
    runes := []rune(input)
    for i, j := 0, len(runes)-1; i < j; i, j = i+1, j-1 {
        runes[i], runes[j] = runes[j], runes[i]
    }
    return string(runes)
}

func shortestPalindrome(s string) string {
    revS := reverse(s)

    for i:=0; i<len(s); i++ {
        ls := s[:len(s)-i]
        rs := revS[i:len(s)]
        if ls == rs {
            return revS + s[len(s)-i:]
        }
    }

    return s
}
